File: src/rna2fuun/rna2fuun.py
# -*- coding: utf-8 -*-

import sys
import Image
import ImageColor
import ImageDraw
import functools

    
sys.setrecursionlimit(600*600)

mode = 'RGBA'
size = (600, 600)
black = (0 , 0 , 0)
red = (255, 0 , 0)
green = (0 , 255, 0)
yellow = (255, 255, 0)
blue = (0 , 0 , 255)
magenta = (255, 0 , 255)
cyan = (0, 255, 255)
white = (255, 255, 255)
transparent = 0
opaque = 255

commands = {
    'PIPIIIC' : "addColor, black",
    'PIPIIIP' : "addColor, red",
    'PIPIICC' : "addColor, green",
    'PIPIICF' : "addColor, yellow",
    'PIPIICP' : "addColor, blue",
    'PIPIIFC' : "addColor, magenta",
    'PIPIIFF' : "addColor, cyan",
    'PIPIIPC' : "addColor, white",
    'PIPIIPF' : "addColor, transparent",
    'PIPIIPP' : "addColor, opaque",
    'PIIPICP' : "emptyBucket",
    'PIIIIIP' : "move",
    'PCCCCCP' : "turnCCW",
    'PFFFFFP' : "turnCW",
    'PCCIFFP' : "mark",
    'PFFICCP' : "line",
    'PIIPIIP' : "tryfill",
    'PCCPFFP' : "addBitmap",
    'PFFPCCP' : "compose",
    'PFFICCF' : "clip"}

E, S, W, N = 0, 1, 2, 3

dstr = {E:"East", S:"South", W:"West", N:"North"}
def d2str(d):
    return dstr[d]

def c2p(c, alpha):
    a = (alpha,)
    return c + a

def splitrna(rna):
    for i in xrange(0, len(rna), 7):
        yield rna[i:i+7] 

def read(filename):
    f = open(filename)
    rna = list(splitrna(f.read()))
    f.close()
    return rna

class rna2fuun(object):

    def __init__(self):
        self.debug = False
        self.reset()

    def reset(self):
        self.bucket = []
        self.position = (0, 0)
        self.mark = (0, 0)
        self.dir = E
        self.bitmaps = []
        self.addBitmap()
        # Cache
        self.avgcol = None
        self.avgalpha = None
        self.currpix = None
    
    def addColor(self, c):
        if self.debug: 
            print "Adding colour:", c
        self.bucket.insert(0, c)
        self.avgcol = None
        self.avgalpha = None
        self.currpix = None
    
    def emptyBucket(self):
        if self.debug: 
            print "Emptying bucket"
        self.bucket = []
        self.avgcol = None
        self.avgalpha = None
        self.currpix = None
    
    def averageColor(self, default):
        if not self.avgcol:
            rc = 0
            gc = 0
            bc = 0
            colbucket = [c for c in self.bucket if isinstance(c, tuple)]
            if colbucket:
                for c in colbucket:
                    r, g, b = c
                    rc += r 
                    gc += g 
                    bc += b
                lencb = len(colbucket)
                self.avgcol = (rc/lencb, gc/lencb, bc/lencb)
            else:
                self.avgcol = (default, default, default)
        return self.avgcol

    def averageAlpha(self, default):
        if not self.avgalpha:
            abucket = [c for c in self.bucket if isinstance(c, int)]
            if abucket:
                self.avgalpha = sum(abucket)/len(abucket)
            else:
                self.avgalpha = default
        return self.avgalpha
    
    def currentPixel(self):
        if not self.currpix:
            rc, gc, bc = self.averageColor(0)
            ac = self.averageAlpha(255)
            self.currpix = (rc*ac/255, gc*ac/255, bc*ac/255, ac)
        return self.currpix

    def move(self):
        if self.debug: 
            print "Move: ", self.position, d2str(self.dir)
        x, y = self.position
        if self.dir == N:
            self.position = (x , (y - 1) % 600)
        elif self.dir == E:
            self.position = ((x + 1) % 600, y )
        elif self.dir == S:
            self.position = (x , (y + 1) % 600)
        elif self.dir == W:
            self.position = ((x - 1) % 600, y )
        else:
            raise "Unknown direction"

    def turnCounterClockwise(self):
        if self.debug: 
            print "Turn CCW: ", d2str(self.dir)
        self.dir = (self.dir - 1) % 4

    def turnClockwise(self):
        if self.debug: 
            print "Turn CW: ", d2str(self.dir)
        self.dir = (self.dir + 1) % 4

    def getPixel(self, p):
        return self.bitmaps[0][1][p]

    def setPixel(self, p):
        self.bitmaps[0][1][p] = self.currentPixel()

    def setPixelVal(self, p, val):
        self.bitmaps[0][1][p] = val

    def linePil(self, start, stop):
        ld = self.bitmaps[0][2].line(start+stop)

    def line(self, start, stop):
        if self.debug: 
            print "Line:", start, stop
        x0, y0 = start
        x1, y1 = stop
        deltax = x1 - x0
        deltay = y1 - y0
        d = max(abs(deltax), abs(deltay))
        
        if (deltax * deltay) >= 0:
            c = 1 
        else:
            c = 0
            
        x = x0 * d + (d - c)/2 
        y = y0 * d + (d - c)/2
        cp = self.currentPixel() 
        for i in xrange(d):
            self.setPixelVal((x/d, y/d), cp)
            x = x + deltax
            y = y + deltay
        self.setPixelVal((x1, y1), cp)
    
    def fill(self, pos, oldp, newp):
        stack = []
        stack.append(pos)
        bm = self.bitmaps[0][1]
        while stack:
            x, y = stack.pop()
            try:
                if bm[x,y] == oldp: 
                    bm[x,y] = newp
                    stack.append((x, y + 1))
                    stack.append((x, y - 1))
                    stack.append((x + 1, y))
                    stack.append((x - 1, y))
            except IndexError:
                pass
    
    def tryfill(self):
        newp = self.currentPixel()
        oldp = self.getPixel(self.position)
        if self.debug: 
            print "Fill: ", self.position, oldp, newp
        if newp != oldp:
            self.fill(self.position, oldp, newp)
        
    def addBitmap(self):
        if self.debug: 
            print "Add bitmap"
        if len(self.bitmaps) < 10:
            b = Image.new(mode, size, c2p(black, transparent))
            self.bitmaps.insert(0, (b, b.load(), ImageDraw.Draw(b)))
        
    def compose(self):
        if self.debug: 
            print "Compose"
        if len(self.bitmaps) >= 2:
            bm0 = self.bitmaps[0][1]
            bm1 = self.bitmaps[1][1]
            for y in xrange(600):
                for x in xrange(600):
                    r0, g0, b0, a0 = bm0[x,y]
                    r1, g1, b1, a1 = bm1[x,y]
                    bm1[x, y] = (r0 + (r1 * (255 - a0) / 255),
                                g0 + (g1 * (255 - a0) / 255),
                                b0 + (b1 * (255 - a0) / 255),
                                a0 + (a1 * (255 - a0) / 255))
            self.bitmaps.pop(0)
        
    def clip(self):
        if self.debug: 
            print "Clip"
        if len(self.bitmaps) > 2:
            bm0 = self.bitmaps[0][1]
            bm1 = self.bitmaps[1][1]
            for y in xrange(600):
                for x in xrange(600):
                    r0, g0, b0, a0 = bm0[x,y]
                    r1, g1, b1, a1 = bm1[x,y]
                    bm1[x, y] = (r1 * a0 / 255,
                                g1 * a0 / 255,
                                b1 * a0 / 255,
                                a1 * a0 / 255)
            self.bitmaps.pop(0)

    def doMark(self):
        if self.debug: 
            print "Mark:", self.position
        self.mark = self.position
            
    def doLine(self):
        self.line(self.position, self.mark)
##        self.linePil(self.position+self.mark, self.currentPixel())

    def buildgenerator(self, rna):
        d = {
            'PIPIIIC' : functools.partial(self.addColor, black),
            'PIPIIIP' : functools.partial(self.addColor, red),
            'PIPIICC' : functools.partial(self.addColor, green),
            'PIPIICF' : functools.partial(self.addColor, yellow),
            'PIPIICP' : functools.partial(self.addColor, blue),
            'PIPIIFC' : functools.partial(self.addColor, magenta),
            'PIPIIFF' : functools.partial(self.addColor, cyan),
            'PIPIIPC' : functools.partial(self.addColor, white),
            'PIPIIPF' : functools.partial(self.addColor, transparent),
            'PIPIIPP' : functools.partial(self.addColor, opaque),
            'PIIPICP' : self.emptyBucket,
            'PIIIIIP' : self.move,
            'PCCCCCP' : self.turnCounterClockwise,
            'PFFFFFP' : self.turnClockwise,
            'PCCIFFP' : self.doMark,
            'PFFICCP' : self.doLine,
            'PIIPIIP' : self.tryfill,
            'PCCPFFP' : self.addBitmap,
            'PFFPCCP' : self.compose,
            'PFFICCF' : self.clip}
        
        for r in rna:
            try:
                d[r]()
                yield r
            except KeyError:
                if self.debug:
                    print "Unkown instruction:", r
                yield
                pass
                
    def save(self, filename):
        self.bitmaps[0][0].save(filename, "PNG")
    
    def build(self, filename):   
        rna = read(filename) 
        d = {
            'PIPIIIC' : functools.partial(self.addColor, black),
            'PIPIIIP' : functools.partial(self.addColor, red),
            'PIPIICC' : functools.partial(self.addColor, green),
            'PIPIICF' : functools.partial(self.addColor, yellow),
            'PIPIICP' : functools.partial(self.addColor, blue),
            'PIPIIFC' : functools.partial(self.addColor, magenta),
            'PIPIIFF' : functools.partial(self.addColor, cyan),
            'PIPIIPC' : functools.partial(self.addColor, white),
            'PIPIIPF' : functools.partial(self.addColor, transparent),
            'PIPIIPP' : functools.partial(self.addColor, opaque),
            'PIIPICP' : self.emptyBucket,
            'PIIIIIP' : self.move,
            'PCCCCCP' : self.turnCounterClockwise,
            'PFFFFFP' : self.turnClockwise,
            'PCCIFFP' : self.doMark,
            'PFFICCP' : self.doLine,
            'PIIPIIP' : self.tryfill,
            'PCCPFFP' : self.addBitmap,
            'PFFPCCP' : self.compose,
            'PFFICCF' : self.clip}
        
        lr = len(rna)
        for ix, r in enumerate(rna):
            pct = ix * 100 / lr
            if pct % 10 == 0:
                print "%02d%%"%pct
            try:
                d[r]()
            except KeyError:
                pass
        self.save(filename+".png")

if __name__=="__main__":
    import psyco
    psyco.full()
    r2f = rna2fuun()
##    r2f.debug = True
    r2f.build(sys.argv[1])

File: src/rna2fuun_c/rna2fuun.cpp
#include "rna2fuun.hpp"

#include <iostream>

const unsigned int w = 600;
const unsigned int h = 600;

const pixel black 	= {0, 0, 0};
const pixel red 	= {255, 0, 0};
const pixel green 	= {0, 255, 0};
const pixel yellow 	= {255, 255, 0};
const pixel blue 	= {0, 0, 255};
const pixel magenta	= {255, 0, 255};
const pixel cyan 	= {0, 255, 255};
const pixel white 	= {255, 255, 255};
const unsigned char transparent = 0;
const unsigned char opaque 		= 255;

const int E = 0;
const int S = 1;
const int W = 2;
const int N = 3;

const int ADDCOLORBLACK = 0;
const int ADDCOLORRED = 1;
const int ADDCOLORGREEN = 2;
const int ADDCOLORYELLOW = 3;
const int ADDCOLORBLUE = 4;
const int ADDCOLORMAGENTA = 5;
const int ADDCOLORCYAN = 6;
const int ADDCOLORWHITE = 7;
const int ADDCOLORTRANSPARENT = 8;
const int ADDCOLOROPAQUE = 9;
const int EMPTYBUCKET = 10;
const int MOVE = 11;
const int TURNCCW = 12;
const int TURNCW = 13;
const int MARK = 14;
const int LINE = 15;
const int TRYFILL = 16;
const int ADDBITMAP = 17;
const int COMPOSE = 18;
const int CLIP = 19;
    
rna2fuun::rna2fuun() 
{ 
	reset();
	m_commands["PIPIIIC"] = ADDCOLORBLACK;
	m_commands["PIPIIIP"] = ADDCOLORRED;
	m_commands["PIPIICC"] = ADDCOLORGREEN;
	m_commands["PIPIICF"] = ADDCOLORYELLOW;
	m_commands["PIPIICP"] = ADDCOLORBLUE;
	m_commands["PIPIIFC"] = ADDCOLORMAGENTA;
	m_commands["PIPIIFF"] = ADDCOLORCYAN;
	m_commands["PIPIIPC"] = ADDCOLORWHITE;
	m_commands["PIPIIPF"] = ADDCOLORTRANSPARENT;
	m_commands["PIPIIPP"] = ADDCOLOROPAQUE;
	m_commands["PIIPICP"] = EMPTYBUCKET;
	m_commands["PIIIIIP"] = MOVE;
	m_commands["PCCCCCP"] = TURNCCW;
	m_commands["PFFFFFP"] = TURNCW;
	m_commands["PCCIFFP"] = MARK;
	m_commands["PFFICCP"] = LINE;
	m_commands["PIIPIIP"] = TRYFILL;
	m_commands["PCCPFFP"] = ADDBITMAP;
	m_commands["PFFPCCP"] = COMPOSE;
	m_commands["PFFICCF"] = CLIP;
}

rna2fuun::~rna2fuun()
{
}

void rna2fuun::reset()
{
	m_cCache = false;
	m_c.r = 0;
	m_c.g = 0;
	m_c.b = 0;
	m_c.a = opaque;
	m_cpCache = false;
	m_cp.r = 0;
	m_cp.g = 0;
	m_cp.b = 0;
	m_cp.a = opaque;
	m_aCache = false;
	m_a = opaque;
    m_x = 0;
    m_y = 0;
    m_mx = 0;
    m_my = 0;
    m_dir = E;
    m_bitmaps.clear();
    addBitmap();
    emptyBucket();
}

void rna2fuun::addColor(const pixel& c)
{
	m_cbucket.push_back(c);
	m_cCache = false;
	m_cpCache = false;
}

void rna2fuun::addAlpha(unsigned char a)
{
	m_abucket.push_back(a);
	m_aCache = false;
	m_cpCache = false;
}

void rna2fuun::emptyBucket()
{
	m_cbucket.clear();
	m_abucket.clear();
	m_cCache = false;
	m_aCache = false;
	m_cpCache = false;
}

pixel rna2fuun::averageColor(unsigned char def) const
{
	if (m_cbucket.empty())
	{
		pixel p;
		p.r = def;
		p.g = def;
		p.b = def;
		p.a = opaque;
		return p;
	}
	else
	{
		if (!m_cCache)
		{
			unsigned int r = 0;
			unsigned int g = 0;
			unsigned int b = 0;
			for (std::vector<pixel>::const_iterator it = m_cbucket.begin(); it != m_cbucket.end(); ++it)
			{
				const pixel& bp = *it;
				r += bp.r;
				g += bp.g;
				b += bp.b;
			}
			size_t sz = m_cbucket.size();
			m_c.r = r / sz;
			m_c.g = g / sz;
			m_c.b = b / sz;
			m_cCache = true;
		}
		return m_c;			
	}
}

unsigned char rna2fuun::averageAlpha(unsigned char def) const
{
	if (m_abucket.empty())
	{
		return def;
	}
	else
	{
		if (!m_aCache)
		{
			unsigned int a = 0;
			for (std::vector<unsigned char>::const_iterator it = m_abucket.begin(); it != m_abucket.end(); ++it)
			{
				a += *it;
			}
			m_a = a / m_abucket.size();
			m_aCache = true;
		}
		return m_a;
	}
}    

const pixel& rna2fuun::currentPixel() const
{
	if (!m_cpCache)
	{
		unsigned char a = averageAlpha(opaque);
		pixel p = averageColor(0);
		m_cp.r = static_cast<unsigned int>(p.r) * static_cast<unsigned int>(a) / 255;		
		m_cp.g = static_cast<unsigned int>(p.g) * static_cast<unsigned int>(a) / 255;		
		m_cp.b = static_cast<unsigned int>(p.b) * static_cast<unsigned int>(a) / 255;
		m_cp.a = a;		
	}
	return m_cp; 
}


void rna2fuun::move()
{
	switch (m_dir)
	{
	case N:
		--m_y;
		if (m_y < 0)
		{
			m_y = h - 1;
		}
		break;
	case E:
		++m_x;
		if (m_x >= w)
		{
			m_x = 0;
		}
		break;
	case S:
		++m_y;
		if (m_y >= h)
		{
			m_y = 0;
		}
		break;
	case W:
		--m_x;
		if (m_x < 0)
		{
			m_x = w - 1;
		}
		break;
	}
}

void rna2fuun::turnCounterClockwise()
{
	--m_dir;
	if (m_dir < 0)
	{
		m_dir = 3;
	}
}

void rna2fuun::turnClockwise()
{
	++m_dir;
	if (m_dir > 3)
	{
		m_dir = 0;
	}
}

void rna2fuun::line()
{
	unsigned int x0 = m_x;
	unsigned int y0 = m_y;
	unsigned int x1 = m_mx;
	unsigned int y1 = m_my;
	
    int deltax = static_cast<int>(x1) - static_cast<int>(x0);
    int deltay = static_cast<int>(y1) - static_cast<int>(y0);
    unsigned int d = std::max(std::abs(deltax), std::abs(deltay));
    
    unsigned int c = 0;
    if ((deltax * deltay) >= 0)
    {
        c = 1;
    }
        
    unsigned int x = x0 * d + (d - c)/2;
    unsigned int y = y0 * d + (d - c)/2;
    pixel cp = currentPixel(); 
    for (unsigned int i = 0; i < d; ++i)
    {
        setPixelVal(x/d, y/d, cp);
        x += deltax;
        y += deltay;
    }
    setPixelVal(x1, y1, cp);
}

void rna2fuun::fill(unsigned int x, unsigned int y, const pixel& oldp, const pixel& newp)
{
	std::vector<std::pair<unsigned int, unsigned int> > stack;
	stack.push_back(std::make_pair(x, y));
    while (!stack.empty())
    {
        std::pair<unsigned int, unsigned int> pos = stack.back();
        stack.pop_back();
        pixel& p = getPixel(pos.first, pos.second);
        if (memcmp(&oldp, &p, sizeof(pixel)) == 0)
        {
        	p = newp;
        	if (pos.second + 1 < h)
        	{
        		stack.push_back(std::make_pair(pos.first, pos.second + 1));
        	}
        	if (pos.first + 1 < w)
        	{
        		stack.push_back(std::make_pair(pos.first + 1, pos.second));
        	}
        	if (pos.second > 1)
        	{
        		stack.push_back(std::make_pair(pos.first, pos.second - 1));
        	}
        	if (pos.first > 1)
        	{
        		stack.push_back(std::make_pair(pos.first - 1, pos.second));
        	}
        }
    }
} 

void rna2fuun::tryfill()
{
	pixel newp = currentPixel();
	pixel oldp = getPixel(m_x, m_y); 
    if (memcmp(&oldp, &newp, sizeof(pixel)) != 0)
    {
        fill(m_x, m_y, oldp, newp);
    }
}

void rna2fuun::addBitmap()
{
	if (m_bitmaps.size() == 10)
	{
		return;
	}
	pixel p;
	p.r = 0;
	p.g = 0;
	p.b = 0;
	p.a = transparent;
	m_bitmaps.insert(m_bitmaps.begin(), std::vector<pixel>(w*h, p));
} 
    
void rna2fuun::compose()
{
    if (m_bitmaps.size() > 1)
    {
        const std::vector<pixel>& bm0 = m_bitmaps[0];
        std::vector<pixel>& bm1 = m_bitmaps[1];
        for (unsigned int i = 0; i < w * h; ++i)
        {
        	const pixel& p0 = bm0[i];
        	pixel& p1 = bm1[i];
            p1.r = p0.r + p1.r * (255 - p0.a) / 255;
            p1.g = p0.g + p1.g * (255 - p0.a) / 255;
            p1.b = p0.b + p1.b * (255 - p0.a) / 255;
            p1.a = p0.a + p1.a * (255 - p0.a) / 255;
        }
        m_bitmaps.erase(m_bitmaps.begin());
    }
}
    
void rna2fuun::clip()
{
    if (m_bitmaps.size() > 1)
    {
        const std::vector<pixel>& bm0 = m_bitmaps[0];
        std::vector<pixel>& bm1 = m_bitmaps[1];
        for (unsigned int i = 0; i < w * h; ++i)
        {
        	const pixel& p0 = bm0[i];
        	pixel& p1 = bm1[i];
            p1.r = p1.r * p0.a / 255;
            p1.g = p1.g * p0.a / 255;
            p1.b = p1.b * p0.a / 255;
            p1.a = p1.a * p0.a / 255;
        }
        m_bitmaps.erase(m_bitmaps.begin());
    }
}    

pixel& rna2fuun::getPixel(unsigned int x, unsigned int y)
{
    return m_bitmaps[0][x + w * y];
}

pixelbuffer rna2fuun::getImage() const 
{ 
	pixelbuffer p;
	p.size = sizeof(pixel) * m_bitmaps[0].size();
	p.data = (const char*)&m_bitmaps[0][0];
	return p;
}


void rna2fuun::mark()
{
	m_mx = m_x;
	m_my = m_y;
}            

void rna2fuun::buildsteps(const std::vector<std::string>& rna)
{
	std::for_each(rna.begin(), rna.end(), std::bind1st(std::mem_fun(&rna2fuun::buildstep), this)); 
}
        
void rna2fuun::buildstep(std::string rna)
{
	std::map<std::string, unsigned int>::const_iterator cit = m_commands.find(rna);
	if (cit != m_commands.end())
	{
		switch (cit->second)
		{
			case ADDCOLORBLACK:
				addColor(black);
				break;
			case ADDCOLORRED:
				addColor(red);
				break;
			case ADDCOLORGREEN:
				addColor(green);
				break;
			case ADDCOLORYELLOW:
				addColor(yellow);
				break;
			case ADDCOLORBLUE:
				addColor(blue);
				break;
			case ADDCOLORMAGENTA:
				addColor(magenta);
				break;
			case ADDCOLORCYAN:
				addColor(cyan);
				break;
			case ADDCOLORWHITE:
				addColor(white);
				break;
			case ADDCOLORTRANSPARENT:
				addAlpha(transparent);
				break;
			case ADDCOLOROPAQUE:
				addAlpha(opaque);
				break;
			case EMPTYBUCKET:
				emptyBucket();
				break;
			case MOVE:
				move();
				break;
			case TURNCCW:
				turnCounterClockwise();
				break;
			case TURNCW:
				turnClockwise();
				break;
			case MARK:
				mark();
				break;
			case LINE:
				line();
				break;
			case TRYFILL:
				tryfill();
				break;
			case ADDBITMAP:
				addBitmap();
				break;
			case COMPOSE:
				compose();
				break;
			case CLIP:
				clip();
				break;
		}
	}
}
